### ISSUE TITLE
Stack memory exhaustion while dumping RomFS content from titles with a very high file entry count.

### ISSUE THREAD

**Issue by linktlh** [DarkMatterCore/nxdumptool#34]
I was interested in this tool as it looks like it could be a good replacement to guiding users through the tedious process of hactool.

When dumping from a BOTW (NA v0.0.0 cart) + Update (1.6.0) it processes a large chunk of Actor/Pack folder (4,847 actorpacks of the 7,203) 
The last file processed seems to be `Npc_oasis009.bactorpack` the file is not correctly dumped and that is when it appears to crash.

here's an atmosphere crash report if that'd be helpful.
[01565712611_010000000000100d.log](https://github.com/DarkMatterCore/nxdumptool/files/3497885/01565712611_010000000000100d.log)

it seems to be very consistent. If running from a game homebrew the Nintendo HOS handles it fairly well and closes the application. However, if launched from the gallery HOS crashes outright (resulting in the crash log above)

Let me know if there's anything else I could provide that'd help out.
Extra Info: SanDisk 64gb [10mb/s] SDXC formated in 2 32gb fat32 partitions

**Comment by DarkMatterCore** [comment]
Thanks for reporting this problem. I'll take a look at the logfile to try to pinpoint the crash location in the codebase.

If it isn't too much to ask, could you please provide with additional testing? Thanks in advance.

**Comment by linktlh** [comment]
Further testing:
Attempting to dump JUST the Npc_oasis009.sbactorpack is impossible as it does not show up in `romfs/Actor/Pack/` folder which is extremely strange... 
Afaik it's not a file added in an update, even so, I have include updates open checked.

Attempting to dump the entire `romfs/Actor/Pack` folder:
The switch does crash when dumping the entire directory. Looks to be the exact same spot as well. 

Is there any way to enable a verbose mode that outputs to a log file?

**Comment by linktlh** [comment]
With some extra testing;
Dumping every directory except Actor (one by one):
Successful on the smaller folders, but crashed on every directory with more than 5k files. 
(Actor, Model, and Terrain)
The only exception to this is Voice (7,562 files)

The consistency makes me want to say an overflow error, but since Voice didn't crash I'm not really sure.
Definitely a strange issue.

**Comment by DarkMatterCore** [comment]
Please test again using the attached build, and provide new logs if possible. This one adds some small fixes and uses latest libnx.

[nxdumptool_test_20190813.zip](https://github.com/DarkMatterCore/nxdumptool/files/3499385/nxdumptool_test_20190813.zip)

**Comment by DarkMatterCore** [comment]
@linktlh RomFS entries are read and displayed as-is, using their internal order. Directories are displayed first because they're stored in a different table.

The B button cancel mechanism needs to be improved. I'll look into it as well.

**Comment by linktlh** [comment]
Ah, so it's just the way Nintendo sorts it. It's just fairly unique to see people sort by Ascii value instead of by Alphabet, Although, since that's how Nintendo does it, there'd be no real reason to change it. Theoretically speaking you'd want to dump all of the contents of the game for modding purposes anyways instead of searching for a specific file to edit.
Still, I love the idea that you can go grab a very specific file in romfs.

First thing; This build definitely speeds through the Actorpacks much quicker than the last build.
As for the crashing, unfortunately, it does still crash, in the exact same spot as before. 
[01565744741_010000000000100d.log](https://github.com/DarkMatterCore/nxdumptool/files/3499426/01565744741_010000000000100d.log)

`Npc_oasis009.sbactorpack` is the last file processed (and still incorrectly processed) 

I'm not sure if they'll be helpful, but I attached the Correctly dumped actorpack, and the improperly dumped pack; 
[Npc_oasis009.zip](https://github.com/DarkMatterCore/nxdumptool/files/3499472/Npc_oasis009.zip)
The actorpacks can be opened with [sarc](https://pypi.org/project/sarc/) and the one file not in physics can be viewed via [aamp](https://pypi.org/project/aamp/)

It's clear that parts of the file are corrupt when comparing them on a hex scale but it may just be a result of being partly written as the software crashes.

**Comment by DarkMatterCore** [comment]
Thanks for all your input in this matter. It's something that definitely needs more testing, though. I'm still trying to find the exact source of this problem.

Here's a new build with some more changes, including the improved cancel button mechanism - it should work by just holding down B for 2 seconds in any operation. Also, a message telling you to wait until the output directory is fully deleted will be displayed if the process fails or if it's cancelled - this is already the default behaviour, but no message about it was being displayed.

I'll be waiting for your input. Thanks again.

[nxdumptool_test_20190814.zip](https://github.com/DarkMatterCore/nxdumptool/files/3500027/nxdumptool_test_20190814.zip)

**Comment by linktlh** [comment]
Seems that github had a goof; That file is not available for download.

**Comment by ninetalescommander** [comment]
I can confirm this is an issue not just for Zelda Breath of the Wild, it seems to be for any game that is large in size. I tried to dump Pokken Tournament DX and Monster Hunter Generations Ultimate and it still crashed and had to go into the ROMFS options directly and take things piece by piece.
I still think this program is the best out there for getting models but this issue is quite glaring.

**Comment by DarkMatterCore** [comment]
@linktlh That's probably my Internet connection's fault, sorry. I have to go through a slow and tedious process just to upload a small file like that. I'll try to reupload the binary as soon as possible.

@ninetalescommander Thanks for reporting this. I have successfully tested this feature in the past with Super Mario Odyssey (owned) and Super Smash Bros. Ultimate (borrowed), as well as some small/free eShop titles I have installed. I never had the chance to test it personally with anything with a very high file count.

**Comment by DarkMatterCore** [comment]
[nxdumptool_test_20190814.zip](https://github.com/DarkMatterCore/nxdumptool/files/3503060/nxdumptool_test_20190814.zip)

Please let me know if this one works.

**Comment by linktlh** [comment]
Sadly same result. The B button does function as intended though! 👍 

The log looks pretty much the same doubt it's helpful at this point but here ya go;
[01565813325_010000000000100d.log](https://github.com/DarkMatterCore/nxdumptool/files/3503170/01565813325_010000000000100d.log)

**Comment by DarkMatterCore** [comment]
After an in-depth analysis of the stack trace from the crash log, this is the best I was able to come up with:

```
Stack Trace:

nxdumptool + 0x2cab0 -> D:/GitHub/nxdumptool/source/dumper.c:4317 -> printProgressBar() call in callrecursiveDumpRomFsFile()
nxdumptool + 0x12188 -> D:/GitHub/nxdumptool/source/util.c:2910   -> uiFill() call in printProgressBar()
nxdumptool + 0x15f4c -> D:/GitHub/nxdumptool/source/ui.c:576      -> FT_Render_Glyph() call in uiFill()
nxdumptool + 0x138dd4 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/base/ftobjs.c:4518
nxdumptool + 0x182f98 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftsmooth.c:321
nxdumptool + 0x182390 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:1934
nxdumptool + 0x18204c -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:1795
nxdumptool + 0x181ed0 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:1718
nxdumptool + 0x139564 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/base/ftoutln.c:196
nxdumptool + 0x182c3c -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:1219

LR -> nxdumptool + 0x182bc8 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:1058
PC -> nxdumptool + 0x182420 -> /home/davem/projects/devkitpro/pacman-packages/switch/freetype/src/freetype-2.9/src/smooth/ftgrays.c:847
```

I'll have to perform additional tests to find out what's actually happening -  but so far, it seems like something funny is going on with FreeType.

**Comment by DarkMatterCore** [comment]
After some additional tests, I was able to confirm that this bug is most likely related to stack memory exhaustion due to excesive usage of code recursion in RomFS-related functions.

A friend of mine owns the game and he has kindly helped me out by providing with private testing. My Internet connection has been worse than terrible these last few days, so I had to find another way to keep working on this.

Even after making an internal build with `stb_truetype` as the font renderization library instead of FreeType, the crash could still be reproduced, albeit at an entirely different place in code that bears no direct relation to RomFS-related functions - just like the offsets from the logs uploaded to this issue so far. This made me suspect we're actually dealing with stack exhaustion.

I'll make some changes in order to upload a new test build ASAP. Yet again, thanks a lot!

**Comment by DarkMatterCore** [comment]
[nxdumptool_test_20190815.zip](https://github.com/DarkMatterCore/nxdumptool/files/3507703/nxdumptool_test_20190815.zip)

Here's it. Took me a while to get it uploaded.

@linktlh @ninetalescommander If it isn't too much to ask, could you guys please test this build with the games you mentioned before? I wanna know if the issue is gone for good.

Thanks in advance.

**Comment by linktlh** [comment]
Success! 🎉 
I'll wait on @ninetalescommander before closing the issue. However, it appears to be resolved in that build.

Only thing I noticed was how slow it is; although I'm doubting there's much you can really do about that. took about 1.5 hours to dump the 14gib BOTW romfs. The upside is it's much easier than teaching someone how to use CLI, and hactool.

**Comment by DarkMatterCore** [comment]
@linktlh Thanks! Glad to know it. And don't worry, I'll take care of closing this issue as soon as I push the changes - nonetheless, I'll also wait until @ninetalescommander tests the latest build.

Sadly, there's not much I can do about the transfer speed. Consider that the application has to deal with manually decrypting RomFS data blocks from the Program NCA, and the process can take a bit more time when dealing with update data - BKTR sections are decrypted and parsed to patch RomFS content (also decrypted) on-the-fly.

Dealing with titles that have a very high file entry count doesn't help as well. Just like transfering a huge directory full of very small files in a PC, the process can be slow.

Maybe overclocking the console a bit while transfering data could help, but that's something I have not yet checked. After all, the "boost mode" in loading screens from BotW's latest update is nothing more than a small overclock.

**Comment by linktlh** [comment]
Glad after much trial and error it seems to have gotten resolved :D 
Ninetales successfully dumped Monster Hunters Gen ultimate and Pokken Tournament.

I'm gonna go ahead and run a hasher on all of the files just to confirm nothing went wrong; but there seems to be the correct amount of files in the botw full romfs dump.

**Comment by ninetalescommander** [comment]
Finished dumping Pokken Tournament DX, Monster Hunter Generations Ultimate, and Zelda Breath of the Wild with no issues.
Thanks for the fix!

**Comment by DarkMatterCore** [comment]
Nah, thank **you** for noticing this problem and creating the issue. You're the real heroes. Otherwise, I wouldn't have noticed it! :smile:

Time for a bugfix release.

**Comment by DarkMatterCore** [comment]
Fixed in ed45a863faa0524d9701dc6a87bc1f51209430eb.
